2011-07-13  Andreas Krebbel  <Andreas.Krebbel@de.ibm.com>

	* sysdeps/unix/sysv/linux/s390/s390-32/sysdep.h
	(INTERNAL_VSYSCALL_NCS): Use r10 for backing up the return address
	register in order to avoid conflicts with the soft frame pointer
	being held in r11 when necessary.
	* sysdeps/unix/sysv/linux/s390/s390-64/sysdep.h
	(INTERNAL_VSYSCALL_NCS): Likewise.

diff --git a/sysdeps/unix/sysv/linux/s390/s390-32/sysdep.h b/sysdeps/unix/sysv/linux/s390/s390-32/sysdep.h
index 5bfdc44..13ce9ab 100644
--- a/sysdeps/unix/sysv/linux/s390/s390-32/sysdep.h
+++ b/sysdeps/unix/sysv/linux/s390/s390-32/sysdep.h
@@ -368,12 +368,12 @@
     DECLARGS_##nr(args)							      \
     register long _ret asm("2");						      \
     asm volatile (							      \
-    "lr 11,14\n\t"							      \
+    "lr 10,14\n\t"                                                           \
     "basr 14,%1\n\t"							      \
-    "lr 14,11\n\t"							      \
+    "lr 14,10\n\t"                                                           \
     : "=d" (_ret)							      \
     : "d" (fn) ASMFMT_##nr						      \
-    : "cc", "memory", "0", "1", "11" CLOBBER_##nr);			      \
+    : "cc", "memory", "0", "1", "10" CLOBBER_##nr);                          \
     _ret; })
 
 /* Pointer mangling support.  */
diff --git a/sysdeps/unix/sysv/linux/s390/s390-64/sysdep.h b/sysdeps/unix/sysv/linux/s390/s390-64/sysdep.h
index 6491a3b..4d45417 100644
--- a/sysdeps/unix/sysv/linux/s390/s390-64/sysdep.h
+++ b/sysdeps/unix/sysv/linux/s390/s390-64/sysdep.h
@@ -370,12 +370,12 @@
     DECLARGS_##nr(args)							      \
     register long _ret asm("2");					      \
     asm volatile (							      \
-    "lgr 11,14\n\t"							      \
+    "lgr 10,14\n\t"                                                          \
     "basr 14,%1\n\t"							      \
-    "lgr 14,11\n\t"							      \
+    "lgr 14,10\n\t"                                                          \
     : "=d" (_ret)							      \
     : "a" (fn) ASMFMT_##nr						      \
-    : "cc", "memory", "0", "1", "11" CLOBBER_##nr);			      \
+    : "cc", "memory", "0", "1", "10" CLOBBER_##nr);                          \
     _ret; })
 
 /* Pointer mangling support.  */
